services/wfm/src/main/java/org/openkilda/wfm/topology/flow/service/FlowService.java

```diff
@@ -0,0 +1,377 @@
+import java.util.Collection;
+import java.util.List;
+
+@Slf4j
+public class FlowService {
+    private TransactionManager transactionManager;
+    private FlowRepository flowRepository;
+    private FlowSegmentRepository flowSegmentRepository;
+    private IslRepository islRepository;
+    private PathComputerFactory pathComputerFactory;
+    private FlowResourcesManager flowResourcesManager;
+
+    public FlowService(PersistenceManager persistenceManager, PathComputerFactory pathComputerFactory,
+                       FlowResourcesManager flowResourcesManager) {
+        transactionManager = persistenceManager.getTransactionManager();
+        RepositoryFactory repositoryFactory = persistenceManager.getRepositoryFactory();
+        flowRepository = repositoryFactory.createFlowRepository();
+        flowSegmentRepository = repositoryFactory.createFlowSegmentRepository();
+        islRepository = repositoryFactory.createIslRepository();
+        this.pathComputerFactory = requireNonNull(pathComputerFactory);
+        this.flowResourcesManager = requireNonNull(flowResourcesManager);
+    }
+
+    public boolean doesFlowExist(String flowId) {
+        return flowRepository.doesExistById(flowId);
+    }
+
+    public FlowPair getFlowPair(String flowId) {
+        Collection<FlowPair> pairs = flowRepository.findFlowPairById(flowId);
+        return pairs.isEmpty() ? null : pairs.iterator().next();
+    }
+
+    public Collection<FlowPair> getFlows() {
+        return flowRepository.findAllFlowPairs();
+    }
+
+    /**
+     * Creates a flow by allocating a path and resources. Stores the flow entities into DB, and
+     * invokes flow rules installation via the command sender.
+     *
+     * @param flow   the flow to be created.
+     * @param sender the command sender for flow rules installation.
+     * @return the created flow with the path and resources set.
+     */
+    public FlowPair createFlow(Flow flow, FlowCommandSender sender)
+            throws RecoverableException, UnroutableFlowException {
+        transactionManager.begin();
+        try {
+            // TODO: the strategy is defined either per flow or system-wide.
+            PathComputer pathComputer = pathComputerFactory.getPathComputer(Strategy.COST);
+            PathPair pathPair = pathComputer.getPath(flow);
+
+            log.info("Creating flow {}. Found path: {}.", flow, pathPair);
+
+            FlowPair flowPair = flowResourcesManager.allocateFlow(buildFlowPair(flow, pathPair));
+
+            flowRepository.createOrUpdate(flowPair);
+
+            final List<FlowSegment> forwardSegments = createFlowSegments(flowPair.getForward());
+            final List<FlowSegment> reverseSegments = createFlowSegments(flowPair.getReverse());
+
+            transactionManager.commit();
+
+            // To avoid race condition in DB updates, we should send commands only after DB transaction commit.
+            sender.sendInstallRulesCommand(flowPair.getForward(), forwardSegments);
```

Comment:
If exception will be raised here... we will try to rollback closed transaction.

Reply:
very nice catch